backend-beeyond/src/main/java/at/htl/beeyond/resource/ApplicationResource.kt

```diff
@@ -120,36 +147,72 @@ class ApplicationResource {
     @Transactional
     fun stopApplication(@PathParam("id") id: Long?): Response? {
         val application = Application.findById<Application>(id)
-                ?: return Response.status(404).build()
+            ?: return Response.status(404).build()
 
-        if(application.status == ApplicationStatus.RUNNING){
+        if (application.status == ApplicationStatus.RUNNING) {
+            finishStopApplication(application, ApplicationStatus.STOPPED)
+            return Response.ok().build()
+        } else {
+            return Response.status(422).entity("Application is not in state "+ApplicationStatus.RUNNING).build()
+        }
+    }
 
-            deploymentService.stop(application)
-            application.status = ApplicationStatus.FINISHED
-            application.finishedAt = LocalDateTime.now()
+    @PATCH
+    @Path("/finish/{id}")
+    @RolesAllowed("teacher")
+    @Transactional
+    fun finishApplication(@PathParam("id") id: Long?): Response? {
+        val application = Application.findById<Application>(id)
+            ?: return Response.status(404).build()
 
-            application.namespace.users.forEach {
-                val notification = Notification(it, "Application has been stopped!", NotificationStatus.NEUTRAL, "application", application.id)
-                notification.persist()
-            }
+        if (application.status == ApplicationStatus.RUNNING || application.status == ApplicationStatus.STOPPED) {
+            finishStopApplication(application, ApplicationStatus.FINISHED)
+            return Response.ok().build()
+        } else {
+            return Response.status(422).entity("Application is not in state "+ApplicationStatus.RUNNING + " or " + ApplicationStatus.STOPPED).build()
+        }
+    }
+
+    @PATCH
+    @Path("/request/{id}")
+    @Transactional
+    fun requestApplication(@PathParam("id") id: Long?): Response? {
+        val application = Application.findById<Application>(id)
+            ?: return Response.status(404).build()
 
-            val isLastApplication = Application
-                    .streamAll<Application>()
-                    .filter {
-                        it.status == ApplicationStatus.RUNNING && it.namespace == application.namespace
-                    }.count() == 0L
+        if (application.status == ApplicationStatus.DENIED) {
+            application.status = ApplicationStatus.PENDING
+            return Response.ok().build()
+        } else {
+            return Response.status(422).entity("Application is not in state " + ApplicationStatus.DENIED).build()
```

Comment:
Application is not in state " + ApplicationStatus.DENIED + " or " + ApplicationStatus. PENDING